docs: add pcg troubleshooting (#1403)
spectrocloud/librarium: content/docs/14-troubleshooting/30-pcg.md

# Private Cloud Gateway (PCG)

> You may encounter one of the following scenarios when deploying and using a Private Cloud Gateway in your data center. Some of the scenarios below apply to any data center, whereas others apply to specific data centers, such as VMware. Each scenario covers a specific problem, including an overview of the issue and its possible causes, followed by a detailed set of debug steps to help identify and resolve the underlying problem.  

Reviewer: ```suggestion
When deploying and using a Private Cloud Gateway in your computing environment, you may encounter one of the following scenarios. Some of the scenarios below apply to all compute environments, whereas others apply to specific self-hosted environments, such as VMware. Each scenario covers a specific problem, including an overview of the issue and its possible causes, followed by a detailed set of debug steps to help identify and resolve the underlying problem.  
```
I touched up the intro and made it clearer that it's not always self-hosted environments.

<br />

1. Initiate the installation in Palette. In this step, you get a pairing code and an installer image. 
2. Deploy the PCG installer in the data center environment.
3. Configure the cloud gateway in Palette, and launch the PCG cluster.

While deploying a PCG, you may encounter one of the following scenarios during the abovementioned steps. Some scenarios below apply to all data center environments, whereas others apply to a specific data center environment, such as VMware. Each scenario covers a specific problem, including an overview, possible causes, and debugging steps.  
<br />

> # Scenairo - Jet Crashback Loop

Reviewer: ```suggestion
# Scenario - Jet Crashback Loop
```

  The **cloud-init-output.log** file will contain error messages if there are failures with connecting to Palette, authenticating, or downloading installation artifacts. A common cause for these errors is incorrect values provided to the OVF template deployment wizard, such as the Palette endpoint or a mistyped pairing code. 
  
  The screenshot below highlights the OVF template properties you must carefully configure and verify before deploying a PCG installer VM.

  ![A screenshot displaying the OVF template properties you  configure while deploying the PCG installer VM](/troubleshooting-pcg-template_properties.png)

3. Double-check the accuracy of the pairing code used for the PCG installer VM. A pairing code is a unique authentication code Palette generates for each PCG installer instance. Confirm that it matches the value you copied from Palette. 

> 4. Ensure the Palette endpoint is correct and has no trailing slash "`/`". If you use Palette SaaS, the default endpoint is `https://console.spectrocloud.com`. If you are using a self-hosted Palette instance, use the domain name as applicable to you. If the Palette endpoint is incorrectly specified, relaunch a new PCG installer VM with the correct values.

Reviewer: ```suggestion
4. Ensure the Palette endpoint is correct and has no trailing slash `/`. If you use Palette SaaS, the default endpoint is `https://console.spectrocloud.com`. If you are using a self-hosted Palette instance, use the domain name as applicable to you. If the Palette endpoint is incorrectly specified, relaunch a new PCG installer VM with the correct values.
```

    - If you choose the DHCP option, enable the Dynamic DNS in your DNS server.

    - Ensure the search domain is correctly defined in the fault domain's DNS settings. The screenshot below displays the VMware cloud properties you configure in Palette. 
    ![A screenshot displaying the VMware cloud properties you configure in Palette](/troubleshooting-pcg-cluster_config_2.png)

7. Suppose the orchestration is stalled due to the incorrect cloud gateway configuration. In that case, you must delete the gateway and reconfigure it again.

8. If the problem persists, download the cluster logs from Palette. The screenshot below will help you locate the button to download logs from the cluster details page. 

  ![A screenshot highlighting how to download cluster logs in from Palette.](/troubleshooting-pcg-download_logs.png)

9. Share the logs with Spectro Cloud's support team at [support@spectrocloud.com](mailto:support@spectrocloud.com).
<br />

# Scenario - No Progress After Creating the Container Manager
Suppose the PCG events in Palette display no progress after the specific event, `Created container manager`. 

> The issue can occur when the gateway installer VM fails to connect to the Spectro Cloud API, [https://api.spectrocloud.com](https://api.spectrocloud.com), to download the installation artifacts. Another potential reason is that the gateway installer may not have the required permissions to store the installation artifacts in the **spectro-templates** folder. The installer downloads the images for the worker nodes and stores them in the **spectro-templates** folder during the cluster provisioning.

Reviewer: ```suggestion
This issue can occur when the gateway installer VM fails to connect to the Palette API endpoint and download the installation artifacts. Another potential reason is that the gateway installer may not have the required permissions to store the installation artifacts in the **spectro-templates** folder. The installer downloads the images for the worker nodes and stores them in the **spectro-templates** folder during the cluster provisioning.
```
We don't have to make the API a markdown URL. If you click on the API URL you'll get an error.

<br />

## Debug Steps
<br />

> 1. Check the network connectivity from the gateway installer VM to the Spectro Cloud API. Ensure that no network issues or firewall restrictions are blocking the connection.

Reviewer: ```suggestion
1. Check the network connectivity from the gateway installer VM to the Palette API endpoint, `https://api.spectrocloud.com`, or your self-hosted Palette's API endpoint. Ensure that no network issues or firewall restrictions are blocking the connection.
```

2. Ensure you have the necessary write permissions for the **spectro-templates** folder. 
<br />

# Scenario - Failed to Deploy Image

Suppose the PCG events in Palette display the `Failed to deploy image: Failed to create govomiClient` error. 

The error can occur if there is a preceding "https://" or "http://" string in the vCenter server URL or the gateway installer VM lacks internet connectivity.
<br />

## Debug Steps
<br />

1. Check the data center server URL in the gateway cluster configurations. Ensure the URL does not include the preceding "http://" or "https://" string. 

2. Check the internet connectivity from within the gateway installer VM. Ensure that no network issues or firewall restrictions are blocking the connection.

> 3. Suppose the gateway installer VM does not have outbound internet access. In that case, you can either provision the gateway installer VM in a network that supports outbound connectivity or fix the firewall rules of the existing network.

Reviewer: ```suggestion
3. If the gateway installer VM does not have outbound internet access. Either provision the gateway installer VM in a network that supports outbound connectivity, or fix the firewall rules of the existing network to allow outbound internet access for the gateway installer.
```

<br />

> # Scenario - No Route to the Kube API Server

Reviewer: ```suggestion
# Scenario - No Route to the Kubernetes API Server
```

> While deploying the cloud gateway cluster, the Palette events display the `No route to host` error. 

Reviewer: ```suggestion
While deploying the cloud gateway cluster, if events logs display the error message `No route to host.` 
```

> The error indicates an issue with the PCG cluster nodes attempting to connect to the Kube API server. This issue can occur due to improper networking configuration or an error in the cloud-init process. 

Reviewer: ```suggestion
The error indicates an issue with the PCG cluster nodes attempting to connect to the cluster's Kubernetes API server. This issue can occur due to improper networking configuration or an error in the cloud-init process. 
```

<br />

## Debug Steps
<br />

> 1. Ensure you provision the gateway cluster VM in a network supporting outbound connectivity. 

Reviewer: What is the outbound connection targeting? Palette?  I'm not sure what the outbound connection is for.

2. Check the internet connectivity from within the gateway installer VM. Ensure that no network issues or firewall restrictions are blocking the connection.

3. Suppose the gateway installer VM does not have outbound internet access. In that case, you can either provision the gateway installer VM in a network that supports outbound connectivity or fix the firewall rules of the existing network.

> 4. If the gateway cluster VM has a public IP address, SSH into the gateway installer VM using the username `ubuntu` and the SSH key you provided during the OVA import.

Reviewer: ```suggestion
4. If the gateway cluster VM has a public IP address assigned, SSH into the gateway installer VM using the username `ubuntu` and the public SSH key you provided during the OVA import.
```

5. Inspect the log files in the **/var/log** directory. 

 

> 6. Examine the cloud-init and sys logs for potential errors or warnings.

Reviewer: ```suggestion
6. Examine the cloud-init and system logs for potential errors or warnings.
```

<br />

# Scenario - Permission Denied to Provision

> A user must have the necessary permissions to provision a PCG cluster in the VMware data center. 

Reviewer: ```suggestion
You must have the necessary permissions to provision a PCG cluster in the VMware environment. 
```

> Suppose you do not have adequate permissions. In that case, the PCG cluster provisioning will fail, and you will get the "*Permission to perform this operation denied*" error. 

Reviewer: ```suggestion
If you do not have adequate permissions, the PCG cluster provisioning will fail, and you will get the `Permission to perform this operation denied` error message in the event log. 
```

<br />

## Debug Steps
<br />

> 1. Ensure you have all the permissions listed in the [VMware Privileges](https://docs.spectrocloud.com/clusters/data-center/vmware/#vmwareprivileges) section before proceeding to provision a PCG cluster. 

Reviewer: ```suggestion
1. Ensure you have all the permissions listed in the [VMware Privileges](/clusters/data-center/vmware/#vmwareprivileges) section before proceeding to provision a PCG cluster. 
```

> 2. Contact your vCenter administrator if you need permission.

Reviewer: ```suggestion
2. Contact your VMware administrator if you are missing any of the required permissions.
```

6. Check if the deployment failed due to a lack of outbound internet connectivity from the PCG installer VM. Use the following steps to check outbound internet connectivity:
    - SSH into the PCG installer VM using the username `ubuntu` and the public SSH key you provided during the OVA import. 
    - Use the ping command to check if the VM can reach a public IP address. For example, ping well-known public IPs like Google's public DNS server (8.8.8.8) or any other public IP address.
    <br />

      ```bash
      ping 8.8.8.8
      ```
    - If you receive responses from the ping requests, it indicates that the VM has outbound internet connectivity.
    - Suppose you do not receive a response from the ping requests. In that case, go to the next step for further troubleshooting steps. 

7. Check for any network restrictions or firewall rules in the data center's network settings that may block communication. Adjust the proxy settings, if applicable, to fix the connectivity. Alternatively, you can power down and delete the PCG installer VM and relaunch a new one in a network that supports outbound internet connections.

> 8. If the problem persists, email the log files to Spectro Cloud's support team at [support@spectrocloud.com](mailto:support@spectrocloud.com).

Reviewer: ```suggestion
8. If the problem persists, email the log files to our support team at [support@spectrocloud.com](mailto:support@spectrocloud.com).
```

The error indicates an issue with the PCG cluster nodes attempting to connect to the cluster's Kubernetes API server. This issue can occur due to improper networking configuration or an error in the cloud-init process. 
<br />

## Debug Steps
<br />

1. Check the data center network settings. Ensure no network restrictions, firewalls, or security groups block communication between the nodes and the API server.

2. If you use the DHCP allocation scheme, check that the DHCP service is available on the DHCP server. Restart the service if it's not in an active state.

3. If you use the DHCP allocation scheme, ensure Dynamic DNS is enabled in the DHCP server. A Dynamic DNS is only required if you are using DHCP. Dynamic DNS is not required for a static IP allocation scheme.

4. Check the Kubernetes API server status. The Kubernetes API server must be active and healthy on the control plane node. Use the following steps to check the status.

    - Switch to [Palette](https://console.spectrocloud.com).

    - Navigate to the **Tenant Settings** > **Private Cloud Gateways** page. 

    - Click on the newly provisioned PCG cluster to review its details. 

>     - Download the PCG cluster's Kubeconfig file from the **Overview** tab. Click on the Kubeconfig file name to download it to your local machine, as highlighted in the screenshot below. 

Reviewer: ```suggestion
    - Download the PCG cluster's kubeconfig file from the **Overview** tab. Click on the kubeconfig file name to download it to your local machine, as highlighted in the screenshot below. 
```

  
      ![A screenshot highlighting the kubeconfig file to download from Palette.](/troubleshooting-pcg-download_kubeconfig.png)
    

>     - After you download the PCG cluster's Kubeconfig file, use the following commands to make a GET request to one of the [Kubernetes API server endpoints](https://kubernetes.io/docs/reference/using-api/health-checks/#api-endpoints-for-health), `/readyz` or `'/livez'`. Replace `[path_to_kubeconfig]` placeholder with the path to the Kubeconfig file you downloaded in the previous step. A status code `ok` or `200` indicates the Kubernetes API server is healthy. 

Reviewer: ```suggestion
    - After you download the PCG cluster's kubeconfig file, use the following commands to make a GET request to one of the [Kubernetes API server endpoints](https://kubernetes.io/docs/reference/using-api/health-checks/#api-endpoints-for-health), `/readyz` or `'/livez'`. Replace `[path_to_kubeconfig]` placeholder with the path to the kubeconfig file you downloaded in the previous step. A status code `ok` or `200` indicates the Kubernetes API server is healthy. 
```

4. If the problem persists, contact our support team by sending an email to [support@spectrocloud.com](mailto:support@spectrocloud.com).
<br />

# Scenario - IP Address Not Assigned During PCG Creation

> When deploying a PCG in VMware vCenter, the IP address fails to get assigned to the new node. This can be caused by networking errors such as DHCP issues or improper execution of cloud-init.

Reviewer: ```suggestion
When deploying a PCG in VMware vCenter, the IP address fails to get assigned to the new node. This can be caused by networking errors such as DHCP issues or an incomplete cloud-init.
```
Execute is an aggressive term, so we try to avoid it.

<br />

## Debug Steps
<br />

1. Verify the network connectivity between the new node and the DHCP server. Ensure that the DHCP server is operational and has available IP addresses in its pool.

2. Check the network configuration on the new node and ensure that it is set to obtain an IP address automatically via DHCP.

> 3. Review the cloud-init logs on the new node for any errors or warnings related to the IP address assignment process.

Reviewer: As I read each step and put my n00b hat on, I can't help but ask "How do I do this"? How does a user check the connectivity? Is it as simple as us suggesting a `ping` between the instances? How can they verify their network configuration is correct? What file and configuration is expected?

These instructions don't really help the reader, we can do better.

<br />

> # Scenario - Cluster Provisioning Failure: "No Route to Host" to Kube API Server

Reviewer: ```suggestion
# Scenario -Missing Network Route to Kube API Server
```
We can shorten this title.

> While provisioning a cluster, the event shows `no route to host` to the Kube API Server. This issue can occur due to improper networking configuration, errors in the cloud-init logs on the newly created VM, or sys logs indicating errors in the newly created VM.

Reviewer: ```suggestion
While provisioning a cluster, you may encounter an event log message stating `no route to host`.  This indicates an issue when the PCG cluster nodes are attempting to connect to the Kube API Server. This issue can occur due to improper networking configuration, or errors in the cloud-init process in the newly created VM. Use the following steps to resolve the route error.
```

This paragraph was difficult to read and needed touches to make it clearer to the reader what the scenario us.

<br />

## Debug Steps
<br />

1. Review the networking configuration for the cluster and ensure that it is properly set up with the correct IP addresses, subnets, and gateways.

2. Check the cloud-init logs on the newly created VM for any errors or warnings that might indicate a problem with network configuration or initialization.

> 3. Examine the sys logs on the newly created VM for any errors or issues that might be causing network connectivity problems.

Reviewer: Same feedback as above. How can we help the reader? What are the expected configurations?
